pkg/scheduler/internal/queue/pod_backoff.go

```diff
@@ -0,0 +1,119 @@
+	// lock for performing actions on this PodBackoffMap
+	lock sync.Mutex
+	// initial backoff duration
+	initialDuration time.Duration
+	// maximal backoff duration
+	maxDuration time.Duration
+	// map for pod -> number of attempts for this pod
+	podAttempts map[ktypes.NamespacedName]int
+	// map for pod -> lastUpdateTime pod of this pod
+	podLastUpdateTime map[ktypes.NamespacedName]time.Time
+}
+
+// NewPodBackoffMap creates a PodBackoffMap with initial duration and max duration.
+func NewPodBackoffMap(initialDuration, maxDuration time.Duration) *PodBackoffMap {
+	return &PodBackoffMap{
+		initialDuration:   initialDuration,
+		maxDuration:       maxDuration,
+		podAttempts:       make(map[ktypes.NamespacedName]int),
+		podLastUpdateTime: make(map[ktypes.NamespacedName]time.Time),
+	}
+}
+
+// GetBackoffTime returns the time that nsPod completes backoff
+func (pbm *PodBackoffMap) GetBackoffTime(nsPod ktypes.NamespacedName) (time.Time, bool) {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	if _, found := pbm.podAttempts[nsPod]; found == false {
+		return time.Time{}, false
+	}
+	lastUpdateTime := pbm.podLastUpdateTime[nsPod]
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	backoffTime := lastUpdateTime.Add(backoffDuration)
+	return backoffTime, true
+}
+
+// TryBackoffAndWait tries to perform backoff for a non-preempting pod.
+// it is invoked from factory.go if util.PodPriorityEnabled() returns false.
+func (pbm *PodBackoffMap) TryBackoffAndWait(nsPod ktypes.NamespacedName, stop <-chan struct{}) bool {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	select {
+	case <-time.After(backoffDuration):
+		return true
+	case <-stop:
+		return false
+	}
+}
+
+// calculateBackoffDuration is a helper function for calculating the backoffDuration
+// based on the number of attempts the pod has made.
+func (pbm *PodBackoffMap) calculateBackoffDuration(nsPod ktypes.NamespacedName) time.Duration {
+	backoffDuration := pbm.initialDuration
+	if _, found := pbm.podAttempts[nsPod]; found {
+		for i := 1; i < pbm.podAttempts[nsPod]; i++ {
+			backoffDuration = backoffDuration * 2
+			if backoffDuration > pbm.maxDuration {
+				return pbm.maxDuration
+			}
+		}
+	}
+	return backoffDuration
+}
+
+// clearPodBackoff removes all tracking information for nsPod.
+// Lock is supposed to be acquired by caller.
+func (pbm *PodBackoffMap) clearPodBackoff(nsPod ktypes.NamespacedName) {
+	delete(pbm.podAttempts, nsPod)
+	delete(pbm.podLastUpdateTime, nsPod)
+}
+
+// CleanupPodsCompletesBackingoff execute garbage collection on the pod backoff,
+// i.e, it will remove a pod from the PodBackoffMap if
+// lastUpdateTime + maxBackoffDuration is before the current timestamp
+func (pbm *PodBackoffMap) CleanupPodsCompletesBackingoff() {
+	pbm.lock.Lock()
+	for pod, value := range pbm.podLastUpdateTime {
+		if value.Add(pbm.maxDuration).Before(time.Now()) {
+			pbm.clearPodBackoff(pod)
+		}
+	}
+	pbm.lock.Unlock()
```

Comment:
the body of this function is non-trivial. Please use `defer pbm.lock.Unlock()` in the beginning of the function.